### PR TITLE
Feature: tools - Add flatten_directory method

### PR DESCRIPTION
**Pull request by MaciejKucia**
Changelog: (Feature): Added new tools method - flatten_directory
Docs: https://github.com/conan-io/docs/pull/1958

- [ ] Refer to the issue that supports this Pull Request.
- [ ] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

#TAGS: slow

### CODE REVIEW

**Comment by MaciejKucia** [comment]
@memsharded Please advise if anything is needed here. I went through CONTRIBUTING but I am not sure if associated issue is required or not. Thank you!

File: conans/test/unittests/util/tools_test.py [hunk 1]
```diff
@@ -876,6 +876,31 @@ def build(self):
         client.save({"conanfile.py": conanfile, "file.txt": "hello\r\n"})
         client.run("create . user/channel")
 
+    @attr('slow')
+    @pytest.mark.slow
+    @attr('local_bottle')
```
Comment (memsharded) [inline conans/test/unittests/util/tools_test.py:881]: A test with a local bottle running shouldn't be necessary to test this utility, this mark can be removed

File: conans/client/tools/files.py [hunk 2]
```diff
@@ -471,3 +476,32 @@ def work_on_element(dirpath, element, token):
...
+    base = Path(base_path)
+    matching_directories = [d for d in base.iterdir() if d.is_dir() and fnmatch(d.name, pattern)]
+    if not matching_directories:
+        raise ConanException("Failed to flatten directory: No directory found")
+    if len(matching_directories) > 1:
+        raise ConanException("Failed to flatten directory: Only single directory can be selected")
+
+    directory = matching_directories[0]
+    directory_name = directory.name
+    randomized = str(directory) + str(random.random())
```
Comment (memsharded) [inline conans/client/tools/files.py:501]: This is a bit fragile. Maybe it is very unlikely, but a random string can still collide.

File: conans/test/unittests/util/tools_test.py [hunk 3]
```diff
@@ -876,6 +876,31 @@ def build(self):
         client.save({"conanfile.py": conanfile, "file.txt": "hello\r\n"})
         client.run("create . user/channel")
 
+    @attr('slow')
+    @pytest.mark.slow
```
Comment (memsharded) [inline conans/test/unittests/util/tools_test.py:880]: This shouldn't be a slow test at all, this mark can be removed

**Review by memsharded — COMMENTED** [review]
Maybe the main use case happens when unzipping one file from the internet? https://github.com/conan-io/conan/pull/8208 recently added the ``strip_root`` to ``tools.download()`` and ``tools.get()``, ``tools.unzip``, etc. it will be available in Conan 1.33.

Maybe that would be enough, or you are interested in this functionality for other cases?

**Comment by memsharded** [comment]
Also, broken CI https://ci.conan.io/blue/organizations/jenkins/ConanTestSuite/detail/PR-8157/3/pipeline, doesn't seem related to this PR. I'd suggest to merge the develop branch.

**Comment by MaciejKucia** [comment]
Thanks, that recent PR does solve my use-case where I unzip local file.

**Comment by memsharded** [comment]
Thanks @MaciejKucia for contributing and following up!

**Pull request closed without merging**